File: code/samples/python/src/roles/shopping_agent/agent.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""A shopping agent.

The shopping agent's role is to engage with a user to:
1. Find products offered by merchants that fulfills the user's shopping intent.
2. Help complete the purchase of their chosen items.

The Google ADK powers this shopping agent, chosen for its simplicity and
efficiency in developing robust LLM agents.
"""

from common.retrying_llm_agent import RetryingLlmAgent
from common.system_utils import DEBUG_MODE_INSTRUCTIONS

from . import tools
from .subagents.payment_method_collector.agent import payment_method_collector
from .subagents.shipping_address_collector.agent import (
    shipping_address_collector,
)
from .subagents.shopper.agent import shopper


root_agent = RetryingLlmAgent(
    max_retries=5,
    model="gemini-3.1-flash-lite-preview",
    name="root_agent",
    instruction="""
          You are a shopping agent responsible for helping users find and
          purchase products from merchants.

          Follow these instructions, depending upon the scenario:

    %s

          Scenario 1: The user asks to buy or shop for something.
          Execute these steps STRICTLY in order. Do NOT skip steps 1, 2, or 3,
          and do NOT advance to step 6 before step 3 has completed.
          1. Delegate to the `shopper` agent to find products and let the
             user select one.
          2. Delegate to `shipping_address_collector` to get the shipping
             address. Display it to the user.
          3. Delegate to `payment_method_collector` to get the payment method.
             This sub-agent MUST present the eligible payment methods to the
             user and obtain their selection before you may continue.
             Do not invent a payment method; do not proceed without the
             sub-agent completing.
          4. Call `create_checkout` to get the merchant-signed checkout JWT
             for the selected cart.
          5. Present the cart summary (item, price, shipping address, payment
             method) and ask the user to confirm.
          6. When the user confirms, call these tools in order:
             a. `create_checkout_mandate` — creates a CheckoutMandate SD-JWT
             b. `create_payment_mandate` — creates a PaymentMandate SD-JWT
             c. `send_signed_mandates_to_credentials_provider` — this will
                fail if step 3 was skipped, at which point you must go back
                and delegate to `payment_method_collector`.
          7. Call `initiate_payment` to start the payment.
          8. If prompted for an OTP, relay the request to the user. Once they
             provide the OTP, call `initiate_payment_with_otp`.
          9. On success, present the payment receipt to the user.

          CRITICAL: Never display raw tool outputs, JSON responses (like checkout
          data or API messages), or SD-JWT strings to the user. Keep these internal.
          Only show the human-readable summaries like the Cart Summary and the
          Payment Receipt.

          Scenario 2: Anything else.
          Respond: "Hi, I'm your shopping assistant. How can I help you?
          For example, you can say 'I want to buy a pair of shoes'"
          """ % DEBUG_MODE_INSTRUCTIONS,
    tools=[
        tools.create_checkout,
        tools.create_checkout_mandate,
        tools.create_payment_mandate,
        tools.send_signed_mandates_to_credentials_provider,
        tools.initiate_payment,
        tools.initiate_payment_with_otp,
    ],
    sub_agents=[
        shopper,
        shipping_address_collector,
        payment_method_collector,
    ],
)

File: code/samples/typescript/src/roles/shopping-agent/agent.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LlmAgent } from '@google/adk';
import { DEBUG_MODE_INSTRUCTIONS } from '../../common/constants/index.js';

import {
  updateCart,
  createPaymentMandate,
  signMandatesOnUserDevice,
  sendSignedPaymentMandateToCredentialsProvider,
  initiatePayment,
  initiatePaymentWithOtp,
} from './tools.js';

// Import sub_agents
import { shopperAgent } from './subagents/shopper/agent.js';
import { shippingCollectorAgent } from './subagents/shipping-address-collector/agent.js';
import { paymentCollectorAgent } from './subagents/payment-method-collector/agent.js';

/**
 * Shopping Agent (ADK)
 *
 * Main orchestrator for the entire shopping and payment flow.
 * Uses ADK sub_agents for Shopper, ShippingCollector, and PaymentCollector
 * (matching Python's architecture where these share the same process and session state).
 */
export const shoppingAgent = new LlmAgent({
  name: 'root_agent',
  model: 'gemini-3.1-flash-lite',
  description: 'A shopping agent responsible for helping users find and purchase products from merchants.',
  instruction: `You are a shopping agent responsible for helping users find and purchase products from merchants.

You have three subagents that you delegate to:
- shopper_agent: Helps users find and select products from merchants.
- shipping_address_collector_agent: Collects the user's shipping address.
- payment_method_collector_agent: Collects the user's payment method.

When you delegate to a subagent, ADK handles the delegation automatically. Simply describe what you need done and the appropriate subagent will handle it.

Follow these instructions, depending upon the scenario:

Scenario 1:
The user asks to buy or shop for something.
1. Delegate to the shopper_agent to collect the products the user is interested in purchasing. The shopper_agent will return a message indicating if the chosen cart mandate is ready or not.
2. Once a success message is received, delegate to the shipping_address_collector_agent to collect the user's shipping address.
3. The shipping_address_collector_agent will return the user's shipping address. Display the shipping address to the user.
4. Once you have the shipping address, call the updateCart tool to update the cart. You will receive a new, signed CartMandate object.
5. Immediately after the updateCart tool returns successfully, delegate to the payment_method_collector_agent to collect the user's payment method.
6. The payment_method_collector_agent will return the user's payment method alias.
7. Send this message separately to the user: 'This is where you would be redirected to a trusted surface to confirm the purchase.' 'But this is a demo, so you can confirm your purchase here.'
8. Call the createPaymentMandate tool to create a payment mandate.
9. Present to the user the final cart contents using EXACTLY the amounts from the orderSummary in the createPaymentMandate result: each line item (item price, shipping, tax) and the total price, plus how long the cart is valid for (in a human-readable format) and how long it can be refunded (in a human-readable format). Never estimate, recompute, or re-attribute amounts. In a second block, show the shipping address. Format it all nicely. In a third block, show the user's payment method alias. Format it nicely.
10. Confirm with the user they want to purchase the selected item using the selected form of payment.
11. When the user confirms purchase call the following tools in order:
   a. signMandatesOnUserDevice
   b. sendSignedPaymentMandateToCredentialsProvider
12. Initiate the payment by calling the initiatePayment tool.
13. If prompted for an OTP, relay the OTP request to the user. Do not ask the user for anything other than the OTP request. Once you have a challenge response, display the display_text from it and then call the initiatePaymentWithOtp tool to retry the payment. Surface the result to the user.
14. If the response is a success or confirmation, create a block of text titled 'Payment Receipt'. Report the payment status, the payment ID, and EXACTLY the amounts from the orderSummary in the tool result: each line item (item price, shipping, tax) and the total price. Never estimate, recompute, or re-attribute amounts. In a second block, show the shipping address. Format it all nicely. In a third block, show the user's payment method alias. Format it nicely and give it to the user.

Scenario 2:
The users ask you do to anything else.
1. Respond to the user with this message: "Hi, I'm your shopping assistant. How can I help you?  For example, you can say 'I want to buy a pair of shoes'"

CRITICAL: Never display raw tool outputs, JSON responses (like cart data, mandates, or API messages), or SD-JWT strings to the user. Keep these internal. Only show the human-readable summaries like the Cart Summary and the Payment Receipt. Never introduce yourself by your internal agent name and never mention tool names, subagent names, or internal object names (IntentMandate, CartMandate, PaymentMandate) when talking to the user — speak like a friendly shopping assistant. The only exception is when the user has explicitly asked for debug or verbose mode.

${DEBUG_MODE_INSTRUCTIONS}`,
  tools: [
    updateCart,
    createPaymentMandate,
    signMandatesOnUserDevice,
    sendSignedPaymentMandateToCredentialsProvider,
    initiatePayment,
    initiatePaymentWithOtp,
  ],
  subAgents: [
    shopperAgent,
    shippingCollectorAgent,
    paymentCollectorAgent,
  ],
});

// ADK devtools discovers the root agent via a `rootAgent` named export.
export { shoppingAgent as rootAgent };
